### PR TITLE
Revert direct interval edits implementation 

### PR DESCRIPTION
**Pull request by jzaffiro**
Identical to #2, with a cleaner branch. I addressed some of the comments from the previous PR, but will add the outstanding ones here. 

[AB#4011](https://dev.azure.com/fluidframework/internal/_workitems/edit/4011)

### CODE REVIEW

File: packages/dds/sequence/src/revertibles.ts [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+	interval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+	collection: IntervalCollection<SequenceInterval>,
+) {
+	// TODO: create localReferences to track positions of deleted interval
+	// just use positions on the interval?
+	revertibles.push({
+		event: IntervalEventType.DELETE,
+		interval,
+		start: interval.start,
```
Comment (jzaffiro) [inline packages/dds/sequence/src/revertibles.ts]: I'm not sure you can just use the existing local references and expect them to still be valid by the time the revert happens. (Also for changeInterval)
_from previous pr_

File: packages/dds/sequence/src/revertibles.ts [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+
+function revertLocalAdd(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.ADD>,
+) {
+	// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+	const id = revertible.interval.getIntervalId()!;
+	// might be better to assert if id is undefined
+	const label = revertible.interval.properties.label;
+	// should I use revertible.collection here or access from sharedstring?
```
Comment (jzaffiro) [inline packages/dds/sequence/src/revertibles.ts]: > It's better not to have redundancy. If we don't need both, you can remove one.

Not sure what redundancy this refers to. If it is having 2 different ways to access the intervalCollection, is it even possible to remove one of them?
_from previous pr_

File: packages/dds/sequence/src/revertibles.ts [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+
+	return revertibles;
+}
+
+function revertLocalAdd(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.ADD>,
+) {
+	// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+	const id = revertible.interval.getIntervalId()!;
```
Comment (jzaffiro) [inline packages/dds/sequence/src/revertibles.ts]: I think it's true but worth checking with Abram that an interval id is sufficiently reliable across edits to guarantee that it is the interval we think it is.
_from previous pr_

File: packages/dds/sequence/src/revertibles.ts [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+	const end = string.localReferencePositionToPosition(revertible.end);
+
+	revertible.collection.change(id, start, end);
+}
+
+function revertLocalPropertyChanged(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.PROPERTYCHANGED>,
+) {
+	const label = revertible.interval.properties.label;
```
Comment (jzaffiro) [inline packages/dds/sequence/src/revertibles.ts]: We could store the interval collection on the revertible instead.

Which makes me wonder if we should always do that. Like what if the interval collection is deleted and then a new one with the same label is created?
_from previous pr_

File: packages/dds/sequence/src/revertibles.ts (hunk 3 above)
Comment (jzaffiro) [reply packages/dds/sequence/src/revertibles.ts]: @Abe27342

Comment (Abe27342) [reply packages/dds/sequence/src/revertibles.ts]: it is. in fact I think we can actually clean up the fact that this API can return `undefined` as well: it's back-compat.

Basically in 0.39 and earlier, we didn't put guid IDs on intervals as we didn't support a `change` operation. But we have since 0.40, and there's a codepath to assign default non-guid IDs for intervals we encounter in such documents. We should leave that logic around, but since it exists it's the case that all intervals have immutable IDs for the duration of their lifetime.

File: packages/dds/sequence/src/revertibles.ts (hunk 4 above)
Comment (Abe27342) [reply packages/dds/sequence/src/revertibles.ts]: interval collection creation/deletion is implicit--once a collection is materialized in memory, it stays around and is always valid to use. It would get GC'd at a similar timeframe to whenever its associated SharedString does.

We don't bother including empty interval collections in the snapshot, and we don't issue explicit create or delete operations for them.

So it should generally be fine to keep interval collections by reference wherever you want.

File: packages/dds/sequence/src/revertibles.ts (hunk 2 above)
Comment (Abe27342) [reply packages/dds/sequence/src/revertibles.ts]: looking at that comment thread, it should never matter due to my other comment about interval collection lifetime. That said, I agree with being consistent: if it's not overly verbose in code, just holding on to the sharedString and getting the right collection when you need it would be less memory at low runtime cost (cost of getting an interval collection from a SharedString when that collection already exists is basically just a map lookup), so I'd maybe lightly lean slightly toward that.

File: packages/dds/sequence/src/revertibles.ts (hunk 1 above)
Comment (Abe27342) [reply packages/dds/sequence/src/revertibles.ts]: you shouldn't be able to, but I'm glad this got brought up. I think we have a leak here where we're not actually cleaning up local references for deleted/modified intervals. So I wouldn't be surprised if something like this works right now. I'd expect you to need to clone the endpoints here though.

Comment (Abe27342) [reply packages/dds/sequence/src/revertibles.ts]: https://dev.azure.com/fluidframework/internal/_workitems/edit/4065

Comment (scarlettjlee) [reply packages/dds/sequence/src/revertibles.ts]: I couldn't find where they get deleted so I'm also suspicious of a leak. I dunno when a good time to clean those up would be. Maybe when the min seq passes them. Or after the callbacks are invoked, and therefore any listeners should never hang on to them.

This undo-redo code shouldn't be adding to that problem. At the least, any local references created here should get deleted when the revert is invoked. From the app perspective, they probably want to hold on to them for the lifetime of the session. Otherwise maybe we should consider something like a dispose later.

Comment (Abe27342) [reply packages/dds/sequence/src/revertibles.ts]: in the absence of the undo-redo stack we should just be deleting them immediately on operation, no need to wait for minSeq to pass by.

With undo-redo enabled, that would mean we need to clone refs or create ones at the same positions. Agreed that refs should get deleted when the revert is invoked.

File: packages/dds/sequence/src/revertibles.ts [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+	return revertibles;
+}
+
+export function appendLocalDeleteToRevertibles(
+	string: SharedString,
+	interval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+) {
+	const startRef = interval.start;
+	const endRef = interval.end;
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/revertibles.ts]: You should call createLocalReferencePosition to clone these since they may change by the time revert is called. (And removeLocalReferencePosition on revert.)

File: packages/dds/sequence/src/revertibles.ts [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+	return revertibles;
+}
+
+function revertLocalAdd(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.ADD>,
+) {
+	const id = revertible.interval.getIntervalId();
+	const label = revertible.interval.properties.referenceRangeLabels[0];
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/revertibles.ts:143]: Is it safe to use referenceRangeLabels like this? If an interval in a collection is guaranteed to have exactly one, maybe that functionality should move to the interval or interval collection class.

File: api-report/sequence.api.md [hunk 7]
```diff
@@ -318,7 +318,7 @@ export class SequenceInterval implements ISerializableInterval {
     compareEnd(b: SequenceInterval): number;
     compareStart(b: SequenceInterval): number;
     end: LocalReferencePosition;
-    getIntervalId(): string | undefined;
```
Comment (scarlettjlee) [inline api-report/sequence.api.md]: Not a big deal, but this change could go into main instead.

File: packages/dds/sequence/src/revertibles.ts [hunk 8]
```diff
@@ -0,0 +1,203 @@
...
+	string: SharedString,
+	interval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+) {
+	const startSeg = interval.start.getSegment() as SharedStringSegment;
+	const endSeg = interval.end.getSegment() as SharedStringSegment;
+	const startRef = string.createLocalReferencePosition(
+		startSeg,
+		interval.start.getOffset(),
+		ReferenceType.RangeBegin,
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/revertibles.ts]: Doesn't the type need to be | StayOnRemove to prevent them being deleted?

File: packages/dds/sequence/src/test/revertibles.spec.ts [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+		const id = collection.add(0, 5, IntervalType.Simple, { foo: "one" }).getIntervalId()!;
+		collection.changeProperties(id, { foo: "two" });
+
+		revertIntervalRevertibles(sharedString, revertibles.splice(0));
+		assertIntervals(sharedString, collection, [{ start: 0, end: 5 }]);
+		const int = collection.getIntervalById(id);
+		assert.equal(int?.properties.foo, "one");
+	});
+});
+describe("Sequence.Revertibles with Remote Edits", () => {
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/test/revertibles.spec.ts:205]: Remote edit cases should also include cases where the remote edits were interval edits. e.g. local change properties, remote change properties. I can help brainstorm if you want.

File: packages/dds/sequence/src/revertibles.ts [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+}
+// Uses of referenceRangeLabels will be removed once AB#4081 is completed.
+function revertLocalAdd(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.ADD>,
+) {
+	// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+	const id = revertible.interval.getIntervalId()!;
+	const label = revertible.interval.properties.referenceRangeLabels[0];
+	string.getIntervalCollection(label).removeIntervalById(id);
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/revertibles.ts:144]: I think it's true that interval collection will no-op in all the calls if the interval has been [remote] deleted in the meantime, but double check that.

File: packages/dds/sequence/src/revertibles.ts [hunk 11]
```diff
@@ -0,0 +1,203 @@
...
+function revertLocalChange(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.CHANGE>,
+) {
+	const label = revertible.interval.properties.referenceRangeLabels[0];
+	// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+	const id = revertible.interval.getIntervalId()!;
+	const start = string.localReferencePositionToPosition(revertible.start);
+	const end = string.localReferencePositionToPosition(revertible.end);
+	string.getIntervalCollection(label).change(id, start, end);
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/revertibles.ts:178]: It's arguable what should happen if a remote move/change happened in the meantime. E.g. The interval is at the beginning of the string, I moved the interval to the middle of the string, a remote client moves it to the end, I revert my edit. Similarly for property changes. We could handle this, but I'm inclined to keep the code simple the way it is here. What do you think?

Comment (Abe27342) [reply packages/dds/sequence/src/revertibles.ts:178]: overwriting the remote edit is consistent with the analogous conflicts you mentioned for property changes on MergeTree segments (though by "property changes" I guess you could mean on intervals--either way similar case). Keeping it simple is fine imo.

File: packages/dds/sequence/src/revertibles.ts (hunk 10 above)
Comment (jzaffiro) [reply packages/dds/sequence/src/revertibles.ts:144]: If the interval doesn't exist at the time of removal, there will not be an event emitted.

File: packages/dds/sequence/src/test/revertibles.spec.ts [hunk 12]
```diff
@@ -90,11 +101,93 @@ describe("Sequence.Revertibles with Local Edits", () => {
...
+		sharedString.insertText(0, "hello world");
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+		const id = collection.add(0, 5, IntervalType.Simple).getIntervalId()!;
+		collection.change(id, 3, 8);
+		collection.removeIntervalById(id);
+
+		revertIntervalRevertibles(sharedString, revertibles.splice(0));
+		assertIntervals(sharedString, collection, []);
+	});
+	it("performs two local changes, then reverts the first", () => {
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/test/revertibles.spec.ts:151]: This is actually reverting both the add and change, which doesn't match the description unless it means "reverts everything up to the first".

Also, it's not really supported to revert local edits out of order, although I think it would work fine with your code.

File: packages/dds/sequence/src/test/revertibles.spec.ts [hunk 13]
```diff
@@ -90,11 +101,93 @@ describe("Sequence.Revertibles with Local Edits", () => {
...
+
+		sharedString.insertText(0, "hello world");
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+		const id = collection.add(0, 5, IntervalType.Simple).getIntervalId()!;
+		collection.change(id, 3, 8);
+
+		revertIntervalRevertibles(sharedString, revertibles.splice(0));
+		assertIntervals(sharedString, collection, []);
+	});
+	it("performs two local changes, acks the second, and then reverts the first", () => {
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/test/revertibles.spec.ts]: Same comment as previous test.

And acks for a client won't happen out of order. The one message that gets acked is the insertText here.

File: packages/dds/sequence/src/test/revertibles.spec.ts [hunk 14]
```diff
@@ -22,17 +22,29 @@ import { IntervalCollection, IntervalType, SequenceInterval } from "../intervalC
 import { SharedStringFactory } from "../sequenceFactory";
 import { assertIntervals } from "./intervalUtils";
 
-describe("Sequence.Revertibles with Local Edits", () => {
+describe.only("Sequence.Revertibles with Local Edits", () => {
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/test/revertibles.spec.ts]: Don't forget to remove .only here and line 186.

File: packages/dds/sequence/src/test/revertibles.spec.ts [hunk 15]
```diff
@@ -169,14 +283,187 @@ describe("Sequence.Revertibles with Remote Edits", () => {
...
+		const id = collection.add(0, 5, IntervalType.Simple).getIntervalId()!;
+		containerRuntimeFactory.processAllMessages();
+
+		collection.on("deleteInterval", (interval, local, op) => {
+			appendLocalDeleteToRevertibles(sharedString, interval, revertibles);
+		});
+		collection.removeIntervalById(id);
+		containerRuntimeFactory.processAllMessages();
+
+		collection2.change(id, 3, 8);
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/test/revertibles.spec.ts:336]: I think you want to remove the previous processAllMessages, which causes client2 to see the interval remove so the change is a no-op.

Comment (scarlettjlee) [reply packages/dds/sequence/src/test/revertibles.spec.ts:336]: Although it's going to be a no-op on client 1 anyway.

Comment (scarlettjlee) [reply packages/dds/sequence/src/test/revertibles.spec.ts:336]: Maybe the test is more interesting if you swap the order of the remove and change.

File: packages/dds/sequence/src/test/revertibles.spec.ts [hunk 16]
```diff
@@ -169,14 +283,187 @@ describe("Sequence.Revertibles with Remote Edits", () => {
...
+		const id = collection.add(0, 5, IntervalType.Simple).getIntervalId()!;
+		containerRuntimeFactory.processAllMessages();
+
+		collection.on("deleteInterval", (interval, local, op) => {
+			appendLocalDeleteToRevertibles(sharedString, interval, revertibles);
+		});
+		collection.removeIntervalById(id);
+		containerRuntimeFactory.processOneMessage();
+
+		collection2.change(id, 3, 8);
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/test/revertibles.spec.ts]: Same problem as the other test with the same title. Also update the title so they're different (ack or none before the revert).

File: packages/dds/sequence/src/test/revertibles.spec.ts [hunk 17]
```diff
@@ -169,14 +283,187 @@ describe("Sequence.Revertibles with Remote Edits", () => {
...
+		const id = collection.add(0, 5, IntervalType.Simple).getIntervalId()!;
+		containerRuntimeFactory.processAllMessages();
+
+		collection.on("deleteInterval", (interval, local, op) => {
+			appendLocalDeleteToRevertibles(sharedString, interval, revertibles);
+		});
+		collection.removeIntervalById(id);
+		containerRuntimeFactory.processAllMessages();
+
+		collection2.removeIntervalById(id);
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/test/revertibles.spec.ts:417]: Reverse the order of processAllMessages and remote delete so that the remote delete does something on client 2. With this order, client 2 see the client 1 delete before it locally deletes.

File: packages/dds/sequence/src/test/revertibles.spec.ts [hunk 18]
```diff
@@ -169,14 +283,187 @@ describe("Sequence.Revertibles with Remote Edits", () => {
...
+		const id = collection.add(0, 5, IntervalType.Simple, { foo: "one" }).getIntervalId()!;
+		containerRuntimeFactory.processAllMessages();
+
+		collection.on("deleteInterval", (interval, local, op) => {
+			appendLocalDeleteToRevertibles(sharedString, interval, revertibles);
+		});
+		collection.removeIntervalById(id);
+		containerRuntimeFactory.processAllMessages();
+
+		collection2.changeProperties(id, { foo: "two" });
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/test/revertibles.spec.ts:436]: Same thing here: client 2 has already seen the remove. Just get rid of the processAllMessages right before this. It might maybe be more interesting to have the server sequence the changeProperties before the remove. Either way, the remote property change gets thrown away, I think, which is fine.

File: packages/dds/sequence/src/test/revertibles.spec.ts [hunk 19]
```diff
@@ -90,11 +101,93 @@ describe("Sequence.Revertibles with Local Edits", () => {
...
+
+		collection.change(id, 3, 8);
+		containerRuntimeFactory.processOneMessage();
+
+		revertIntervalRevertibles(sharedString, revertibles.splice(0));
+		assertIntervals(sharedString, collection, []);
+	});
 });
-describe("Sequence.Revertibles with Remote Edits", () => {
+describe.only("Sequence.Revertibles with Remote Edits", () => {
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/test/revertibles.spec.ts]: I think your cases so far have avoided hitting this, but in the case where the local and remote edits are both of the same kind that you're subscribing to the events on, you'll have to filter to only add to revertibles for local edits.

File: packages/dds/sequence/src/test/revertibles.spec.ts [hunk 20]
```diff
@@ -169,14 +283,187 @@ describe("Sequence.Revertibles with Remote Edits", () => {
...
+		containerRuntimeFactory.processAllMessages();
+
+		revertIntervalRevertibles(sharedString, revertibles.splice(0));
+		containerRuntimeFactory.processAllMessages();
+
+		assertIntervals(sharedString, collection, [{ start: 0, end: 5 }]);
+		assertIntervals(sharedString2, collection2, [{ start: 0, end: 5 }]);
+		const int = collection.getIntervalById(id);
+		assert.equal(int?.properties.foo, "two");
 	});
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/test/revertibles.spec.ts:488]: Add a case where there are local and remote property changes, with a property that conflicts and one that doesn't.

File: packages/dds/sequence/src/test/revertibles.spec.ts (hunk 12 above)
Comment (scarlettjlee) [reply packages/dds/sequence/src/test/revertibles.spec.ts:151]: Never mind. There's no revertible for the change since only add is subscribed to.

File: packages/dds/sequence/src/test/revertibles.spec.ts (hunk 13 above)
Comment (scarlettjlee) [reply packages/dds/sequence/src/test/revertibles.spec.ts]: This test can just be removed.

File: packages/dds/sequence/src/test/revertibles.spec.ts [hunk 21]
```diff
@@ -300,20 +285,40 @@ describe.only("Sequence.Revertibles with Remote Edits", () => {
 		const id = collection.add(0, 5, IntervalType.Simple).getIntervalId()!;
 		containerRuntimeFactory.processAllMessages();
 
+		collection2.change(id, 3, 8);
+		containerRuntimeFactory.processAllMessages();
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/test/revertibles.spec.ts]: I don't think this and the next test are necessary. I don't think there's anything conflicting between the two clients because client 2's change is seen by both clients before client 1 does a revertible edit + revert. So it's the same as revertible delete + revert over any baseline (in this case, there's agreement among clients that the interval is at [3,8) before any revertible changes) which is mostly covered by the local tests above. You can either get rid of the processAllMessages after the client 2 edit so client 1 doesn't see it until after its local delete, or turn these into a single client test with ack (if that case isn't already covered above).

File: packages/dds/sequence/src/test/revertibles.spec.ts (hunk 18 above)
Comment (scarlettjlee) [reply packages/dds/sequence/src/test/revertibles.spec.ts:436]: I meant to remove the processAllMessages above, not below. The way it is now, `collection2.changeProperties(id, { foo: "two" });` doesn't do anything because client 2 has already applied the interval remove from client 1.

File: packages/dds/sequence/src/test/revertibles.spec.ts [hunk 22]
```diff
@@ -0,0 +1,508 @@
...
+
+		revertIntervalRevertibles(sharedString, revertibles.splice(0));
+		containerRuntimeFactory.processAllMessages();
+
+		assert.equal(sharedString.getText(), sharedString2.getText());
+		assertIntervals(sharedString, collection, [{ start: 0, end: 10 }]);
+		assertIntervals(sharedString2, collection2, [{ start: 0, end: 10 }]);
+	});
+	it("remote interval change interacting with reverting an interval remove", () => {
```
Comment (jzaffiro) [inline packages/dds/sequence/src/test/revertibles.spec.ts:313]: this test still fails with the 0x2f5 assertion, even after updating revertLocalDelete to use the setSlideOnRemove code to set the type correctly. Is there something else I have to do, or anything else that could cause this problem? @Abe27342

Comment (scarlettjlee) [reply packages/dds/sequence/src/test/revertibles.spec.ts:313]: The problem is that the remote change is being applied over the locally re-created interval that is not yet acked, so the endpoints are unexpectedly stayOnRemove. Not sure what the right fix is. We might have to drop remote edits to pending re-creates. Maybe re-create with a new id if it's not valid to hang on to ids across undo-redo.

File: packages/dds/sequence/src/test/revertibles.spec.ts (hunk 14 above)
Comment (Abe27342) [reply packages/dds/sequence/src/test/revertibles.spec.ts]: fyi--we have CI validation for this at runtime in common mocha config. so shouldn't be possible to check in anything that executes a `.only`.

File: packages/dds/sequence/src/test/revertibles.spec.ts (hunk 22 above)
Comment (Abe27342) [reply packages/dds/sequence/src/test/revertibles.spec.ts:313]: recreating with a new id seems like a safe fix. *maybe* we want to support stable id across undo-redo at some point, but i'm not sure anyone cares right now.

File: packages/dds/sequence/src/intervalCollection.ts [hunk 23]
```diff
@@ -1663,6 +1663,7 @@ export class IntervalCollection<TInterval extends ISerializableInterval> extends
 		intervalType: IntervalType,
 		props?: PropertySet,
 	): TInterval {
+		assert(intervalType === IntervalType.SlideOnRemove, "interval type should be slide");
```
Comment (jzaffiro) [inline packages/dds/sequence/src/intervalCollection.ts]: A couple of SharedIntervalCollection tests fail on this assert - should I leave it in this PR or make a separate PR in next to make this change? I could also remove the type parameter all together if it goes into next.

Comment (scarlettjlee) [reply packages/dds/sequence/src/intervalCollection.ts]: It's not related to the changes in this PR. I'd leave it out.

File: packages/dds/sequence/src/revertibles.ts [hunk 24]
```diff
@@ -147,7 +147,7 @@ function revertLocalDelete(
 	const end = string.localReferencePositionToPosition(revertible.end);
 	const intType = revertible.interval.intervalType & ~IntervalType.Simple;
 	const type = intType | IntervalType.SlideOnRemove;
-	const props = revertible.interval.properties;
+	const { intervalId, ...props } = revertible.interval.properties;
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/revertibles.ts:156]: Add a comment here that it's necessary to strip out the id so that the revert creates a new interval id rather than reusing which leads to eventual consistency bugs.

File: packages/dds/sequence/src/test/revertibles.spec.ts [hunk 25]
```diff
@@ -320,6 +320,26 @@ describe.only("Sequence.Revertibles with Remote Edits", () => {
...
 		assertIntervals(sharedString2, collection2, [{ start: 0, end: 5 }]);
 	});
+	it("remote interval change interacting with reverting an interval remove with ack before remove", () => {
+		sharedString.insertText(0, "hello world");
+		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+		const id = collection.add(0, 5, IntervalType.SlideOnRemove).getIntervalId()!;
+		containerRuntimeFactory.processAllMessages();
+
+		collection2.change(id, 3, 8);
+		containerRuntimeFactory.processAllMessages();
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/test/revertibles.spec.ts]: Because of the processAllMessages, I don't think there's any difference between this and the interval being created at 3,8 in the first place. I'm not sure what this test is trying to test.

File: packages/dds/sequence/src/revertibles.ts [hunk 26]
```diff
@@ -0,0 +1,205 @@
...
+
+function revertLocalDelete(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.DELETE>,
+) {
+	const label = revertible.interval.properties.referenceRangeLabels[0];
+	const start = string.localReferencePositionToPosition(revertible.start);
+	const end = string.localReferencePositionToPosition(revertible.end);
+	const intType = revertible.interval.intervalType & ~IntervalType.Simple;
+	const type = intType | IntervalType.SlideOnRemove;
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/revertibles.ts]: Change the type back to simply revertible.interval.intervalType.

File: packages/dds/sequence/src/test/revertibles.spec.ts (hunk 22 above)
Comment (scarlettjlee) [reply packages/dds/sequence/src/test/revertibles.spec.ts:313]: Uh oh: if the id changes, prior revertibles won't find the interval anymore. E.g. change an interval, delete it, revert the delete, then try to revert the change -> change revert no-ops because it doesn't find the interval.
~~The revertibles hold on the intervals themselves rather than the id, so I think it should work if `revertLocalDelete` updates the id on the interval in the revertible.~~ <- this doesn't work because the interval gets recreated (on change, tree update, ack). We could hold a static mapping of ids, or ignore changes to pending deletes.

(Also add a test for this.)

File: packages/dds/sequence/src/test/revertibles.spec.ts (hunk 25 above)
Comment (jzaffiro) [reply packages/dds/sequence/src/test/revertibles.spec.ts]: Removed this test, as it would have been the same as the test above it without the call to processAlMessages after the remote edit. I noticed a few other tests with this problem as well, so I removed the process calls after the remote edits in those.

File: packages/dds/sequence/src/revertibles.ts [hunk 27]
```diff
@@ -0,0 +1,205 @@
...
+	interval: SequenceInterval,
+	revertibles: IntervalRevertible[],
+) {
+	const startSeg = interval.start.getSegment() as SharedStringSegment;
+	const endSeg = interval.end.getSegment() as SharedStringSegment;
+	const startRef = string.createLocalReferencePosition(
+		startSeg,
+		interval.start.getOffset(),
+		ReferenceType.StayOnRemove | ReferenceType.RangeBegin,
+		startSeg.properties,
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/revertibles.ts]: I maybe understand the comment you put on my PR about this now. This local ref is only used by the revertible, so it doesn't need and shouldn't have properties. Even RangeBegin and RangeEnd aren't necessary, although I don't think they're harmful. (Same for all the local references created for revertibles.)

File: packages/dds/sequence/src/revertibles.ts [hunk 28]
```diff
@@ -16,6 +16,8 @@ const IntervalEventType = {
 	PROPERTYCHANGED: 3,
 } as const;
 
+export const idMap = new Map<string, string>();
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/revertibles.ts]: Don't export this. The important thing is to test that the outward behaviour is correct where revertibles function correctly even when an interval has been deleted and undone. E.g. move interval, delete interval, undo delete, undo move. (For which you still need to read the map in the revert functions.)

File: packages/dds/sequence/src/revertibles.ts [hunk 29]
```diff
@@ -0,0 +1,214 @@
...
+	const start = string.localReferencePositionToPosition(revertible.start);
+	const end = string.localReferencePositionToPosition(revertible.end);
+	const type = revertible.interval.intervalType;
+	// reusing the id causes eventual consistency bugs, so it is removed here and recreated in add
+	const { intervalId, ...props } = revertible.interval.properties;
+	const int = string.getIntervalCollection(label).add(start, end, type, props);
+
+	idMap.forEach((oldId, newId) => {
+		if (int.getIntervalId() === oldId) {
+			idMap.set(intervalId, newId);
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/revertibles.ts]: This will never be true since `int` is newly created here. I think you mean:

```suggestion
		if (intervalId === newId) {
			idMap.set(oldId, int.getIntervalId());
```

Also add a test for this. E.g. move interval, remove interval, revert only the remove, remove interval, revert last remove + move.

File: packages/dds/sequence/src/revertibles.ts [hunk 30]
```diff
@@ -0,0 +1,214 @@
...
+
+	return revertibles;
+}
+// Uses of referenceRangeLabels will be removed once AB#4081 is completed.
+function revertLocalAdd(
+	string: SharedString,
+	revertible: TypedRevertible<typeof IntervalEventType.ADD>,
+) {
+	const maybeId = revertible.interval.getIntervalId();
+	const id = idMap.get(maybeId) ?? maybeId;
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/revertibles.ts]: Make this a function. And it also needs to be called in `revertLocalDelete`.

File: packages/dds/sequence/src/test/revertibles.spec.ts [hunk 31]
```diff
@@ -0,0 +1,511 @@
...
+
+		collection2.changeProperties(id, { foo: "two" });
+		containerRuntimeFactory.processAllMessages();
+
+		revertIntervalRevertibles(sharedString, revertibles.splice(0));
+		containerRuntimeFactory.processAllMessages();
+
+		assertIntervals(sharedString, collection, [{ start: 0, end: 5 }]);
+		assertIntervals(sharedString2, collection2, [{ start: 0, end: 5 }]);
+		const int = collection.findOverlappingIntervals(0, 5);
```
Comment (jzaffiro) [inline packages/dds/sequence/src/test/revertibles.spec.ts]: This test fails if I use getIntervalById here instead of findOverlappingIntervals to get the interval. Is that a problem? I expected the new test I added on line 180 to have the same issue, but it doesn't, and I'm confused as to why.

File: packages/dds/sequence/src/revertibles.ts (hunk 6 above)
Comment (scarlettjlee) [reply packages/dds/sequence/src/revertibles.ts:143]: Fix later. AB#4081

File: packages/dds/sequence/src/test/revertibles.spec.ts [hunk 32]
```diff
@@ -174,6 +174,28 @@ describe("Sequence.Revertibles with Local Edits", () => {
 		collection.change(id, 3, 8);
 		collection.removeIntervalById(id);
 
+		revertIntervalRevertibles(sharedString, revertibles.splice(0));
+		assertIntervals(sharedString, collection, [{ start: 0, end: 5 }]);
+	});
+	it("checks id consistency when a remove is undone and redone", () => {
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/test/revertibles.spec.ts]: Nit: this description is slightly misleading since a real "redo" would create a revertible from the add caused by the revert. In this case it's actually more like "checks revertibles still find correct interval across multiple remove and reverts"

File: packages/dds/sequence/src/revertibles.ts [hunk 33]
```diff
@@ -45,6 +45,11 @@ export type IntervalRevertible =
 
 type TypedRevertible<T extends IntervalRevertible["event"]> = IntervalRevertible & { event: T };
 
+export function getId(interval: SequenceInterval): string {
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/revertibles.ts]: Don't export this. It's an internal implementation detail that may change.

File: packages/dds/sequence/src/revertibles.ts [hunk 34]
```diff
@@ -0,0 +1,216 @@
...
+	// reusing the id causes eventual consistency bugs, so it is removed here and recreated in add
+	const { intervalId, ...props } = revertible.interval.properties;
+	const int = string.getIntervalCollection(label).add(start, end, type, props);
+
+	idMap.forEach((oldId, newId) => {
+		if (intervalId === newId) {
+			idMap.set(oldId, getId(int));
+		}
+	});
+	idMap.set(intervalId, getId(int));
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/revertibles.ts]: Whoops, sorry. I told you to call getId in this function, but I meant for re-creating the interval. However, that actually makes no sense. Change it back to what you had before since it will never be in the map.

```suggestion
	idMap.set(intervalId, int.getIntervalId());
```

File: packages/dds/sequence/src/test/revertibles.spec.ts (hunk 31 above)
Comment (scarlettjlee) [reply packages/dds/sequence/src/test/revertibles.spec.ts]: What id are you using? At this point `id` is no longer valid since the interval has been recreated with a new id in the revert. The test on line 180 doesn't call getIntervalId; does it fail if you add that call? It should.

You can use the iterator on `collection` instead of findOverlappingIntervals.

Comment (jzaffiro) [reply packages/dds/sequence/src/test/revertibles.spec.ts]: I was using the id I got back from calling getIntervalId on the initially created interval. The test on 180 doesn't fail if I call getIntervalId, or getIntervalById using the old id value. I'll use the collection iterator in both cases though, just to be safe.

File: packages/dds/sequence/src/test/revertibles.spec.ts [hunk 35]
```diff
@@ -0,0 +1,511 @@
...
+			appendLocalChangeToRevertibles(sharedString, interval, previousInterval, revertibles);
+		});
+
+		collection.change(id, 3, 8);
+		collection.removeIntervalById(id);
+
+		revertIntervalRevertibles(sharedString, revertibles.splice(1, 1));
+		assertIntervals(sharedString, collection, [{ start: 3, end: 8 }]);
+
+		collection.removeIntervalById(id);
```
Comment (scarlettjlee) [inline packages/dds/sequence/src/test/revertibles.spec.ts]: This shouldn't do anything since the id has changed. You'll need to get and use the new id.

File: packages/dds/sequence/src/test/revertibles.spec.ts (hunk 31 above)
Comment (scarlettjlee) [reply packages/dds/sequence/src/test/revertibles.spec.ts]: An interval with the old id should never be found in the collection after it's removed. Can you debug that and see what's going on?

Comment (jzaffiro) [reply packages/dds/sequence/src/test/revertibles.spec.ts]: It looks like the interval isn't found, but we don't fail when removeIntervalById returns undefined. There's also already an entry in revertibles, since only the remove edit is undone the first time, so the test just undoes the change that's in revertibles and doesn't fail.

**Review by scarlettjlee — APPROVED** [review]
